smart_open/tests/test_smart_open.py

```diff
@@ -200,92 +201,173 @@ def test_http_pass(self):
+    class IOReadWrapper(object):
+        def __init__(self):
+            self.fobj = six.BytesIO()
 
-        with open(test_file, 'rb') as infile:
-            compressed_data = infile.read()
+        def read(self, size=-1):
+            return self.fobj.read(size)
 
-        if os.path.isfile(test_file):
-            os.unlink(test_file)
+        def __enter__(self):
+            return self
 
-        responses.add(responses.GET, "http://127.0.0.1/data.bz2", body=compressed_data, stream=True)
-        smart_open_object = smart_open.smart_open("http://127.0.0.1/data.bz2")
+        def __exit__(self, exc_type, exc_val, exc_tb):
+            self.close()
 
-        # decompress the bzip2 and get the same md5 hash
-        self.assertEqual(smart_open_object.read(), test_string)
+        def close(self):
+            self.fobj.close()
 
-    @responses.activate
-    def test_http_xz(self):
-        """Can open xz via http?"""
-        test_string = b'Hello World Compressed.'
-        #
-        # TODO: why are these tests writing to temporary files?  We can do the
-        # lzma compression in memory.
-        #
-        with tempfile.NamedTemporaryFile('wb', suffix='.xz', delete=False) as infile:
-            test_file = infile.name
-
-        with smart_open.smart_open(test_file, 'wb') as outfile:
-            outfile.write(test_string)
+    class IOWriteWrapper(object):
+        def __init__(self):
+            self.fobj = six.BytesIO()
 
-        with open(test_file, 'rb') as infile:
-            compressed_data = infile.read()
+        def write(self, b):
+            return self.fobj.write(b)
 
-        if os.path.isfile(test_file):
-            os.unlink(test_file)
+        def close(self):
+            self.fobj.close()
 
-        responses.add(responses.GET, "http://127.0.0.1/data.xz", body=compressed_data, stream=True)
-        smart_open_object = smart_open.smart_open("http://127.0.0.1/data.xz")
+    def test_read(self):
+        file = self.IOReadWrapper()
+        text = 'Hello, world!'
+        file.fobj.write(text.encode())
+        file.fobj.seek(0)
+        with smart_open.smart_open(file, 'r') as sf:
+            data = sf.read()
+        self.assertEqual(data, text)
+        file = self.IOWriteWrapper()  # no `read()`
+        file.fobj.write(text.encode())
+        file.fobj.seek(0)
+        self.assertRaises(TypeError, lambda: smart_open.smart_open(file, 'r'))
 
-        # decompress the xz and get the same md5 hash
-        self.assertEqual(smart_open_object.read(), test_string)
+    def test_write(self):
+        file = self.IOWriteWrapper()
+        text = 'Hello, world!'
+        with smart_open.smart_open(file, 'w') as sf:
+            sf.write(text)
+            self.assertEqual(file.fobj.getvalue(), text.encode())
+        file = self.IOReadWrapper()  # no `write()`
+        self.assertRaises(TypeError, lambda: smart_open.smart_open(file, 'w'))
+
+    def test_read_write(self):
+        file = six.BytesIO()
```

Comment:
You can avoid the instantiate-write-seek pattern with:

```python
buf = six.BytesIO(text.encode())
```

That will initialize the buffer, fill it with the required contents, and seek to the beginning of the buffer in one call. Please apply this here and elsewhere.